### DISCUSSION TITLE
ORA-01453 "Set transaction must be first statement" while calling sql.DB.BeginTx()

### DISCUSSION THREAD

**Discussion by bobvawter** [godror/godror#334] [Q&A]
I'm occasionally seeing an ORA-01453 error when trying to create a new transaction.  The code base in question also runs against `jackc/pgx`, `go-sql-driver/mysql`, and `go-ora`.

I'm setting up the pool roughly as follows:

```go
params, err := godror.ParseDSN(connectString)
if err != nil {
	return nil, err
}
// Use go's pool, instead of the C library's pool.
params.StandaloneConnection = true
connector := godror.NewConnector(params)
pool := sql.OpenDB(connector)
pool.SetMaxIdleConns(cfg.size)
pool.SetMaxOpenConns(cfg.size)
pool.SetConnMaxIdleTime(cfg.maxIdle)
pool.setConnMaxLifetime(cfg.maxLifetime)
```

Deps:

```
go v1.22.2
github.com/godror/godror v0.42.1
Oracle Instant Client Basic Lite 21.14.0.0.0
```

In general, the pattern that this code base uses is

```
tx := BeginTx()
defer tx.Rollback()
// Do stuff
tx.Commit()
```

There are also plenty of calls to `sql.DB.Exec()` and friends for one-shot statements and queries.

Thanks for any pointers on how I might be setting up the pool incorrectly.

**Comment by tgulacsi** [comment]
Do you use ALTER SESSION / SET TRANSACTION anywhere?

BeginTx DOES use SET TRANSACTION READ WRITE.
This happens iff you doesn't close the (previous) transaction with COMMIT/ROLLBACK.
So, check that every tx is closed properly.

**Reply by bobvawter** [reply]
> Do you use ALTER SESSION / SET TRANSACTION anywhere?

I don't.

> This happens iff you doesn't close the (previous) transaction with COMMIT/ROLLBACK.

If I were to abandon a `sql.Tx` reference, I'd expect to leak the underlying session / network connection, but not to have that connection returned to the pool with a dangling transaction.

This code doesn't call `sql.DB.Conn()`, what other mechanisms would exist that could start a transaction and then return the connection to the pool?

**Reply by tgulacsi** [reply]
You're right, but I've never ever met with ORA-01453 before.
